### PR TITLE
[java] Fix issue #1736 and issue #2207

### LINKED ISSUE #1736: [java] UseStringBufferForStringAppends: False positive if only one concatenation
**Affects PMD Version:** 6.12

**Rule:** [UseStringBufferForStringAppends](https://pmd.github.io/latest/pmd_rules_java_performance.html#usestringbufferforstringappends)

**Description:**
In the code sample provided below PMD wants me to use a StringBuilder to do the concatenation.
That would be the better solution if there were more than one concatenation, especially inside a loop.
But in this case replacing the only concatenation with a StringBuilder would be the same the compiler does anyway.

**Code Sample demonstrating the issue:**

```java
public void doSomething(String s1) {
	String s2 = "a";
	s2 += s1;
}
```

**Running PMD through:** *Eclipse*

### LINKED ISSUE #2207: [java] AvoidInstantiatingObjectsInLoops: False positive - should not flag objects when assigned to lists/arrays
**Affects PMD Version:** 
6.20.0
**Rule:**
[AvoidInstantiatingObjectsInLoops](https://pmd.github.io/latest/pmd_rules_java_performance.html#avoidinstantiatingobjectsinloops)
**Description:**
https://stackoverflow.com/questions/17340421/pmd-avoid-instantiating-new-objects-inside-loops
**Code Sample demonstrating the issue**
From StackOverflow post:

```java
import java.awt.Dimension;

public class PMDDemo {
    public static void main(final String[] args) {
        final Dimension[] arr = new Dimension[10];
        for (int i = 0; i < arr.length; i++) {
            arr[i] = new Dimension(i, i); // rule violation here
        }
    }
}
```

Similarly, if object created is added to a list or collection object.

**Running PMD through:** *[CLI | Ant | Maven | Gradle | Designer | Other]*

Similar issues:
* https://github.com/pmd/pmd/issues/1043#issue-316009471 : calls this case a **false negative**
* https://github.com/pmd/pmd/issues/1615 : wants exception for `new java.util.Date()`

### PR DESCRIPTION
**Pull request by YYoungC**
## Related issues

- Fixes #1736
- Fixes #2207

## Ready?

- [x] Complete build `./mvnw clean verify` passes (checked automatically by travis)
- [x] Added (in-code) documentation (if needed)

### CODE REVIEW

**Comment by pmd-test** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">Running pmdtester failed, this message is mainly used to remind the maintainers of PMD.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

File: pmd-java/src/main/java/net/sourceforge/pmd/lang/java/rule/performance/AvoidInstantiatingObjectsInLoopsRule.java [hunk 1]
```diff
@@ -14,24 +14,77 @@
...
-    private boolean fourthParentNotThrow(ASTAllocationExpression node) {
+    /**
+     * This method is used to check whether the instantiated variable is assigned or not.
+     * @param node This is the expression of part of java code to be checked.
+     * @return boolean This returns Whether the third parent of node is an ASTExpression.
+     */
+    public boolean thirdParentNotASTExpression(ASTAllocationExpression node) {
+        //CS304 Issue link: https://github.com/pmd/pmd/issues/2207
+        if (node.getParent().getClass().toString().equals(
+                "class net.sourceforge.pmd.lang.java.ast.ASTPrimaryPrefix")
```
Comment (adangel) [inline pmd-java/src/main/java/net/sourceforge/pmd/lang/java/rule/performance/AvoidInstantiatingObjectsInLoopsRule.java:43]: Please don't use the `toString()` method for any logic. The `toString()` method is there for debugging purposes and in most cases the format is not an API and can change without any notice. Relying on the format of `toString()` makes java programs brittle. Also, in that case, if we decided to refactor the classes and rename/move ASTPrimaryPrefix, then we won't get here a compile time error but only a runtime error during execution - which is too late.

What you want to achieve here is, to test, whether the parent (or grandparent or greatgrandparent) node is a instance of a specific type:

```java
if (node.getParent() instanceof ASTPrimaryPrefix
    && node.getNthParent(2) instanceof ASTPrimaryExpression) {
    return !(node.getNthParent(3) instanceof ASTExpression);
}
return false;
```

File: pmd-java/src/main/java/net/sourceforge/pmd/lang/java/rule/performance/AvoidInstantiatingObjectsInLoopsRule.java [hunk 2]
```diff
@@ -14,24 +14,77 @@
...
+
+    /**
+     * This method is used to check whether the instantiated variable is assigned or not.
+     * @param node This is the expression of part of java code to be checked.
+     * @return boolean This returns Whether the fourth parent of node is an ASTStatementExpression.
+     */
+    public boolean fourthParentNotASTStatementExpression(ASTAllocationExpression node) {
+        //CS304 Issue link: https://github.com/pmd/pmd/issues/2207
+        return !node.getParent().getParent().getParent().getClass().toString().equals(
+                "class net.sourceforge.pmd.lang.java.ast.ASTStateExpression");
```
Comment (adangel) [inline pmd-java/src/main/java/net/sourceforge/pmd/lang/java/rule/performance/AvoidInstantiatingObjectsInLoopsRule.java:61]: There is no class ASTStateExpression - so this equals always return "true"...
Did you mean "ASTStatementExpression"?
That's by the way one example, why using toString is bad: It didn't show you this error during compile time...

File: pmd-java/src/test/java/net/sourceforge/pmd/lang/java/rule/performance/AvoidInstantiatingObjectsInLoopsTest.java [hunk 3]
```diff
@@ -4,8 +4,98 @@
 
 package net.sourceforge.pmd.lang.java.rule.performance;
 
-import net.sourceforge.pmd.testframework.PmdRuleTst;
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertFalse;
 
-public class AvoidInstantiatingObjectsInLoopsTest extends PmdRuleTst {
```
Comment (adangel) [inline pmd-java/src/test/java/net/sourceforge/pmd/lang/java/rule/performance/AvoidInstantiatingObjectsInLoopsTest.java:9]: You have to keep the base class `PmdRuleTsts` - this provides all the magic to execute the rule test cases. Now you disabled a bunch of tests...
More background info: https://pmd.github.io/latest/pmd_userdocs_extending_testing.html

File: pmd-java/src/test/java/net/sourceforge/pmd/lang/java/rule/performance/AvoidInstantiatingObjectsInLoopsTest.java [hunk 4]
```diff
@@ -4,8 +4,98 @@
...
+    //CS304 (manually written) Issue link: https://github.com/pmd/pmd/issues/2207
+    @Before
+    public void init() {
+        rule = new AvoidInstantiatingObjectsInLoopsRule();
+    }
+
+    //CS304 (manually written) Issue link: https://github.com/pmd/pmd/issues/2207
+    @Test
+    public void visit() {
+        assertEquals("yes", rule.visit(new ASTAllocationExpression(1), "yes"));
```
Comment (adangel) [inline pmd-java/src/test/java/net/sourceforge/pmd/lang/java/rule/performance/AvoidInstantiatingObjectsInLoopsTest.java:27]: I'd say, these tests are too fine granular. They do basically a white box testing - they verify that java works as it should, but don't verify, that the rule does what it should...

Please remove your test cases here and create a real test case: https://pmd.github.io/latest/pmd_userdocs_extending_testing.html

The test cases for AvoidInstantiatingObjectsInLoops can be found here:
https://github.com/pmd/pmd/blob/master/pmd-java/src/test/resources/net/sourceforge/pmd/lang/java/rule/performance/xml/AvoidInstantiatingObjectsInLoops.xml

File: pmd-java/src/main/java/net/sourceforge/pmd/lang/java/rule/performance/UseStringBufferForStringAppendsRule.java [hunk 5]
```diff
@@ -66,13 +77,29 @@ public Object visit(ASTVariableDeclaratorId node, Object data) {
                         ASTAssignmentOperator assignmentOperator = statement
                                 .getFirstDescendantOfType(ASTAssignmentOperator.class);
                         if (assignmentOperator != null && assignmentOperator.isCompound()) {
-                            addViolation(data, assignmentOperator);
+                            //CS304 Issue link: https://github.com/pmd/pmd/issues/1736
+                            // check whether is first time to break the rule
+                            if (!map.containsKey(astName.getNameDeclaration().getName())) {
```
Comment (adangel) [inline pmd-java/src/main/java/net/sourceforge/pmd/lang/java/rule/performance/UseStringBufferForStringAppendsRule.java:82]: Please extract `astName.getNameDeclaration().getName()` into a local variable. You use it 8 times. If I read the code correctly, that's actually the variable's name.

File: pmd-java/src/test/java/net/sourceforge/pmd/lang/java/rule/performance/UseStringBufferForStringAppendsTest.java [hunk 6]
```diff
@@ -4,8 +4,33 @@
...
+    }
+
+    //CS304 (manually written) Issue link: https://github.com/pmd/pmd/issues/1736
+    @Test(expected = NullPointerException.class)
+    public void visit2() {
+        ASTAllocationExpression e = null;
+        assertEquals("yes", rule.visit(e, "yes"));
+    }
 
-public class UseStringBufferForStringAppendsTest extends PmdRuleTst {
```
Comment (adangel) [inline pmd-java/src/test/java/net/sourceforge/pmd/lang/java/rule/performance/UseStringBufferForStringAppendsTest.java:9]: The same comment applies here as well: Please re-enable the rule tests and add additional test cases to https://github.com/pmd/pmd/blob/master/pmd-java/src/test/resources/net/sourceforge/pmd/lang/java/rule/performance/xml/UseStringBufferForStringAppends.xml

File: pmd-java/src/test/java/net/sourceforge/pmd/lang/java/rule/performance/AvoidInstantiatingObjectsInLoopsTest.java (hunk 4 above)
Comment (oowekyala) [reply pmd-java/src/test/java/net/sourceforge/pmd/lang/java/rule/performance/AvoidInstantiatingObjectsInLoopsTest.java:27]: Those tests also use a deprecated constructor and are not compatible with the 7.0 branch. Those constructors should never be called manually.

**Comment by adangel** [comment]
There are still many new violations, but these are ok - they have been actually false-negative, as this rule was not using rule chain previously but still short-circuited the visitor by returning data. Now all AllocationExpressions are visisted, even nested ones in a anonymous class allocation expression, when the anonymous class is not within a loop.

I'll merge that now.

**Pull request merged**